modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java

```diff
@@ -136,14 +158,35 @@ public void onMessage(NetworkMessage message) {
             this.remoteLaunchId = msg.launchId();
             this.remoteConsistentId = msg.consistentId();
 
-            this.recoveryDescriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
+            RecoveryDescriptor descriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
                     remoteConsistentId,
                     remoteLaunchId,
-                    connectionId,
-                    false
+                    connectionId
             );
 
-            handshake(recoveryDescriptor);
+            while (!descriptor.acquire(ctx)) {
+                if (launchId.compareTo(remoteLaunchId) > 0) {
+                    Channel holderChannel = descriptor.holderChannel();
+
+                    if (holderChannel == null) {
+                        continue;
+                    }
+
+                    holderChannel.close().awaitUninterruptibly();
```

Comment:
Is it OK to block the thread? How fast is this operation?
Anyway, uninterruptible awaiting sometimes causes tests to timeout, and you can't kill them by interruption...

Reply:
It is a good question, however in general this should not occur very often, only on first connections. I'd leave it like that for the sake of simplicity. Otherwise we'd need to introduce a future here and handle it in the retry style, which looks horrible